### PR TITLE
Refactoring gen weight storage in EDM + Nano integration

### PR DESCRIPTION
**Pull request by kdlong**
#### PR description:

This PR introduces new weight products to hold the Meta Info of the weights, and the weights themselves, organized into groups of a given type (Scale, Pdf, MEParam, etc.)

This has been presented several times in GEN and X-POG, such as https://indico.cern.ch/event/944653/#5-status-of-new-data-structure

The structure of the products is in place. We are still finalizing the NanoAOD integration and testing for parsing corner cases. We expect these tests to be wrapped up in a couple of weeks, but because of the scale of the changes, it would be useful to have feedback already (I'm sure there will be lost of coding suggestions).

This builds on the work of several people, including @dteague @sroychow @guitargeek @cericeci

@agrohsje and @kpedro88 have contributed to the design and review.

#### PR validation:

We have validated MiniAOD --> NanoAOD, adding products to the event, and GEN-->NanoAOD instantiating products to files. A large testing campaign will take place before NanoAOD v9 and we will report the status here.

#### NanoAOD and MiniAOD file type size changes

The size of the Nano will in general not be changed, by default the same number of PDF and scale weights are stored. The main change is to make the selection of weights to be stored much more convenient, significantly reduce the number of samples where the weights are not found and not stored (makes those individual samples larger but not a big impact on the average), and provides features for keeping more weights in private production.

The change in size of the MiniAOD will depend on the strategy we adopt, which I'd like a bit more feedback on. If we duplicate the event weight information, it is a 10% increase. I assume this is not under consideration. If we zero the weights in the existing product (produce a new LHEEventProduct with only the central weight for example), the change will be negligible. If we only ever produce the product on the fly the size change will be 0.

Perhaps a good option would be to keep the duplicate product at the gen data tier, and zero it out at the MiniAOD step. This requires fixing code that depends on the weights in the LHEEventProduct (Rivet, for example, https://github.com/cms-sw/cmssw/blob/master/GeneratorInterface/RivetInterface/plugins/RivetAnalyzer.cc, but the change is straightforward) and in updating the name of the LHEEventProduct anywhere it is read if it is produced by a new producer (copy previous with zeroed weights). I guess this could be tedious.

### CODE REVIEW

**Comment by kpedro88** [comment]
please test

File: DataFormats/NanoAOD/src/classes_def.xml [hunk 1]
```diff
@@ -9,6 +9,9 @@
     <class name="nanoaod::FlatTable::RowView" transient="true" />
     <class name="edm::Wrapper<nanoaod::FlatTable>" />
 
+    <class name="edm::Wrapper<std::vector<nanoaod::FlatTable>>" />
+    <class name="std::vector<nanoaod::FlatTable>" />
```
Comment (slava77) [inline DataFormats/NanoAOD/src/classes_def.xml:13]: where is this used?
I thought that a table is already a flat vector supporting many types; why make vectors of it

Comment (kdlong) [reply DataFormats/NanoAOD/src/classes_def.xml:13]: This is a long-running discussion in XPOG. An earlier implementation relied on this, and we plan to partially revert back to it. Doing this is part of the WIP tag, but it should come shortly.

Currently we produce one flat table per weight type, to avoid such modifications, but these causes issues for reading single weights from a very large vector branch (~1000 for PDFs with all kept). We want to instead produce a variable number of tables per type (e.g., 10 different PDF tables). We don't know how many tables there will be until parsing the LHE header or reading the products, so we produce a variable number and then flatten the vector at the NanoAOD EDM --> flat stage. the corresponding code for this is in the NanoAODOutputModule/TableOutputbranches code.

**Comment by slava77** [comment]
code-checks

to trigger removal of the reco signature

**Comment by kdlong** [comment]
Would someone be able to trigger Nano/NanoGen tests, as a first check?

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
please test

**Comment by qliphy** [comment]
@kdlong  Any follow up on the unit test failure?

**Comment by kdlong** [comment]
@qliphy I'm going to look closer in the coming days. I can't figure out how to view which tests failed. If I click on details, I see a "Not found" message. Is the output cleared after so many days?

**Comment by kpedro88** [comment]
@kdlong yes, the test output is deleted after some time. We can't start the tests again until the merge conflicts are resolved.

**Comment by kdlong** [comment]
Thanks @kpedro88. I resolved the conflict, if you run the tests again I'll start looking at the failures.

**Comment by kpedro88** [comment]
please test

**Comment by mariadalfonso** [comment]
please test

**Comment by mariadalfonso** [comment]
please test

**Comment by silviodonato** [comment]
@kdlong I converted this PR to draft. Feel free to re-open it as soon as the PR is ready for the review

**Comment by kdlong** [comment]
Hi all,

I'm trying to make a final push to make this converge. We still need to organise a large testing campaign in Gen, but the most recent commits address the outstanding software todo list. The most important software-level change I've made is to remove the dependence on edm::OwnVector, replaced with std::vector<std::unique_ptr<gen::WeightGroupInfo>>.

A couple of additional questions:

1) Do I need to squash commits and/or make a new PR? It's clear that this is quite a monster PR...

2) I've implemented the ability to either produce these products at the GEN step and instantiate them, or to produce them on-the-fly at the Nano step. Producing them at the GEN step is highly desirable, as it allows us to detect and fix issues before a sample is produced, and it also allows a more convenient means of interacting with the weights. An important caveat is that in order to completely replace the weight information in the LHEEventInfoProduct etc, downstream code that depends on it needs to be fixed (such as Rivet). I didn't do this yet, and rather just add the new product in addition to the old product. This duplicates all the weight information, which can be ~1000 floats per event, and therefore the size impact is substantial (~10% in MiniAOD if my measurements are correct). I assume this is not acceptable. So we should decide how to proceed: leave this out of the gen for now (really a shame, as we miss the chance to make improvements for run 3 prod), try to fix the downstream code (could be time consuming), or reduce the number of PDF weights stored for run 3. The latter point I think is a good idea in any case, the vast majority of users don't use them and they have substantial footprint.

**Comment by kdlong** [comment]
Thinking about this a bit more, I think it would be easy to provide a function that returns the weights in the event as a flattened vector, and then to substitute this in any existing code that depends on reading the current product weights formats. Could store a version of the LHEEventInfoProduct with the weights vector containing only the central weight. In principle the GenWeightProduct can be produced along side the LHEEventInfoProduct in externalLHEProducer/LHESource, but I think it's easier to keep the separate producer module as it is.

**Comment by mariadalfonso** [comment]
please test

File: PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py [hunk 2]
```diff
@@ -115,9 +115,10 @@
         'keep *_slimmedGenJetsAK8__*',
         'keep *_slimmedGenJetsAK8SoftDropSubJets__*',
         'keep *_genMetTrue_*_*',
-        # RUN
         'keep LHERunInfoProduct_*_*_*',
         'keep GenRunInfoProduct_*_*_*',
+        'keep GenWeightProduct_*_*_*',
+        'keep GenWeightInfoProduct_*_*_*',
```
Comment (slava77) [inline PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py:121]: is there a way to avoid `*` in the module? in reco we usually require that the module names are spelled out explicitly

**Comment by slava77** [comment]
> 1. Do I need to squash commits and/or make a new PR? It's clear that this is quite a monster PR...

please at least squash. Thank you.

**Comment by slava77** [comment]
@kdlong 
please updated the PR description with expected changes in the file size in mini and in nano.
Thank you.

File: PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py (hunk 2 above)
Comment (kdlong) [reply PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py:121]: Yes, can be

        'keep GenWeightProduct_genWeights_*_*',
        'keep GenWeightInfoProduct_genWeights_*_*',
        'keep GenWeightProduct_lheWeights_*_*',
        'keep GenWeightInfoProduct_lheWeights_*_*',

**Comment by kdlong** [comment]
> @kdlong please updated the PR description with expected changes in the file size in mini and in nano. Thank you.

The size of the Nano will in general not be changed, but it makes the selection of weights to be stored much more convenient, significantly reduces the number of samples where the weights are not found, and provides features for keeping more weights in private production.

The change in size of the MiniAOD will depend on the strategy we adopt, which I'd like a bit more feedback on. If we duplicate the event weight information, it is a 10% increase. I assume this is not under consideration. If we zero the weights in the existing product (produce a new LHEEventProduct with only the central weight for example), the change will be negligible. If we only ever produce the product on the fly the size change will be 0.

I will copy this into the PR description as well.

**Comment by kdlong** [comment]
@slava77 I am trying to come up with a way to squash commits here, but given how long the history is and how interspersed it is with other developments, it is a huge pain. I really don't see a good way to do this that isn't very manual and time consuming (rebase -i would rewrite and enormous amount of history, and if I cherry pick the commits one by one onto master I have to resolve tons of conflicts manually...). Of course I could just checkout a lot of the changes without including the development history, but it'd rather not throw it all away. Do you have any suggestions for this, and is it really necessary?

**Comment by slava77** [comment]
> Of course I could just checkout a lot of the changes without including the development history, but it'd rather not throw it all away.
> Do you have any suggestions for this, and is it really necessary?

I'd suggest to think about how the 270 commits (including intermediate commits that moved files around) are useful for someone looking at this code in the future.
If incremental changes to the same file are not essential, I'd just make a fresh checkout and make (a few) commits of groups of files.

If the history is needed just as a record of growing pains, then perhaps just making a kdlong:GenWeightRefactor_master_forPosterity branch is enough.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by kdlong** [comment]
@slava77 the main thing I wanted to preserve is the authorship. This gives other contributors "credit" for their contributions and also helps maintain it in the future. It's compressed down to 15 commits now which I hope is sufficient.

I've now added the feature to produce a new LHEEventProduct with the weights vector only containing the central weight (though maybe it should just be empty). I think there is still some downstream code to modify, will continue testing next week with help from others. This could also be done for the GenEventInfoProduct, the implementation is the same and I already added it, but the downstream code that reads it is likely much greater, so I haven't done that for now. I'm not sure if it's really worth it, since unlike the LHEEventProduct, the typical size is ~0-45 floats rather than ~1000.

Note that one of the failures in the tests was exactly due to the issue pointed out here: #36705

I've flipped a flag so that it should pass this test now, but I think the correct thing to do is to fix the generators and reset this flag to false, as this failure would help us catch problems in the weight size and metadata in the future.

**Comment by mariadalfonso** [comment]
please test

**Comment by slava77** [comment]
@kdlong 
it looks like the failed test is related to this PR.
Please clarify on the status of resolving the problem.
Thank you.

**Comment by kdlong** [comment]
Hi @slava77 sorry. It seems like we don't handle atypical workflows without the normal gen/lhe weights gracefully, I admittedly never thought to test such things. I'm trying to find time to push this over the line, looking at it together with @sroychow in the coming days

**Comment by mariadalfonso** [comment]
@kdlong @sroychow ping on this PR
if we want to ahem for the Run3 startup we need to have in the 12_4 release

**Comment by mariadalfonso** [comment]
@kdlong 
need to resolve more "code-format" before launching the test

**Comment by kdlong** [comment]
Sorry not sure how that didn't work the first time, I've done it directly with the patch now. I should say that I didn't necessarily address the failures. When I run ```runTheMatrix.py -l 7.3``` it runs successfully. And I don't understand how to do the tau embedding unit test locally. Can someone point me to it?

**Comment by mariadalfonso** [comment]
please test

(previous test got stuck)

**Comment by kdlong** [comment]
I don't get an error when I do ```runTheMatrix.py -l 7.3```. If there is a way I can see the random see used, I can't figure it out. I did manage to track down the tau embedding unit test this time and tested it locally. Setting the switch "failIfInvalidXML = False" fixes this. In principle I would like to keep this true globally and false only for certain workflows. I set another such parameter to a relaxed value as well, at least to test if this fixes the issue in workflow 7.3

**Comment by mariadalfonso** [comment]
please test

**Comment by clacaputo** [comment]
Hi @kdlong about [15ee61d](https://github.com/cms-sw/cmssw/pull/32167/commits/15ee61ddb6fefbefa3857bfedae17b214bfbcf20), did you test the problematic wf `7.3` locally and check if it fix the issue?

**Comment by mariadalfonso** [comment]
please test

**Comment by kdlong** [comment]
Sorry for the repeated failures. What should I do to run every test locally so I don't have to keep spamming this PR?

**Comment by kdlong** [comment]
Or does this imply that I screwed things up that were previously working with my most recent changes? This is possible, let me check.

**Comment by clacaputo** [comment]
> Sorry for the repeated failures. What should I do to run every test locally so I don't have to keep spamming this PR?

You could try `runTheMatrix.py -l limited --ibeos -j 8`

**Comment by clacaputo** [comment]
-reconstruction

**Comment by Saptaparna** [comment]
Hi @kdlong : Indeed, 12_4 will close soon. It would be good to have this PR in there for v10 production.

**Comment by kdlong** [comment]
Hi @Saptaparna: What is the deadline? I gave up on converging quickly some time ago because it seemed too tight, but if there is a realistic amount of time I will see what I can do.

File: GeneratorInterface/Core/src/WeightHelper.cc [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+        return true;
+    }
+    return false;
+  }
+
+  bool WeightHelper::isOrphanPdfWeightGroup(ParsedWeight& weight) const {
+    std::pair<std::string, int> pairLHA;
+    try {
+      pairLHA = LHAPDF::lookupPDF(stoi(searchAttributes("pdf", weight)));
+    } catch (...) {
```
Comment (clacaputo) [inline GeneratorInterface/Core/src/WeightHelper.cc:37]: CMSSW code rules check warns about `catch`
```bash
Rule3 : https://raw.githubusercontent.com/cms-sw/cmssw/master/Utilities/ReleaseScripts/python/cmsCodeRules/config.py
Search for "catch(...)" statements in *.cc, *.cxx files

/GeneratorInterface/Core/src/WeightHelper.cc
[37]
/GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc
[274]
```

In general, do not catch exceptions – leave them to the Framework (see [Exception Guidelines](https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideEdmExceptionUse#Exception%20Handling))

Comment (kpedro88) [reply GeneratorInterface/Core/src/WeightHelper.cc:37]: In this case, we want execution to continue after an exception may be issued by an external package (LHAPDF), so I think catching is appropriate. I think you can put `CMS_SA_ALLOW` before `try` to suppress the warning.

**Comment by mariadalfonso** [comment]
please test

**Comment by sroychow** [comment]
please test

**Comment by sroychow** [comment]
please test

**Comment by swertz** [comment]
@kdlong , would it be possible to rebase on master and make the following changes?

We would propose, to move forward:

- For now, to not run the new products in mini or nano but keep the old ones by default
- Add a customizer to enable the new products and tables in nanoAOD
- Merge this PR in an upcoming pre-release (requires code review)
- Run the test campaign using the customizer
- Once we're happy with the test results, the new products can then be enabled by default and the old ones deprecated

**Comment by swertz** [comment]
enable nano

**Comment by swertz** [comment]
please test

**Comment by swertz** [comment]
Hi @kdlong ,
The failures on the v10 nano workflows are expected if these changes are not compatible with the frozen v10 configuration in https://github.com/cms-sw/cmssw/tree/master/PhysicsTools/NanoAOD/python/V10. However once #40321 is merged this won't be a problem anymore.

About this:
```
An exception of category 'PluginLibraryLoadError' occurred while
   [0] Constructing the EventProcessor
   [1] While attempting to load plugin GenParticles2HepMCConverter
Exception Message:
unable to load /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/32167/29652/CMSSW_13_0_X_2022-12-15-2300/lib/el8_amd64_gcc11/pluginGeneratorInterfaceRivetInterface_plugins.so because /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/32167/29652/CMSSW_13_0_X_2022-12-15-2300/lib/el8_amd64_gcc11/pluginGeneratorInterfaceRivetInterface_plugins.so: cannot open shared object file: Input/output error
```
(on `/TTJets_SingleLeptFromT_TuneCP5_13TeV-madgraphMLM-pythia8/RunIISummer20UL18MiniAODv2-106X_upgrade2018_realistic_v16_L1v1-v2/MINIAODSIM`) it either looks like a missing dependency or a spurious failure, given that a similar sample fails with an entirely different exception below...?

I also see on `/TTToSemiLeptonic_TuneCP5_13p6TeV-powheg-pythia8/Run3Winter22MiniAOD-FlatPU0to70_122X_mcRun3_2021_realistic_v9-v2/MINIAODSIM`:
```
An exception of category 'GenWeightInfoProduct' occurred while
   [0] Processing  Event run: 1 lumi: 1160 event: 814782 stream: 0
   [1] Running path 'dqmoffline_step'
   [2] Prefetching for module NanoAODDQM/'nanoDQMMC'
   [3] Prefetching for module GenWeightsTableProducer/'genWeightsTable'
   [4] Calling method for module GenWeightProductProducer/'genWeightsNano'
Exception Message:
No weight group found containing weight index 46 in the 1 groups.
```

And on `/TTJets_SingleLeptFromT_TuneCP5_13TeV-madgraphMLM-pythia8/RunIISummer20UL17MiniAODv2-106X_mc2017_realistic_v9-v2/MINIAODSIM`:
```
An exception of category 'LHEWeightProductProducer' occurred while
   [0] Processing global begin Run run: 1
   [1] Calling method for module LHEWeightProductProducer/'lheWeightsNano'
Exception Message:
An exception of category 'LHEWeightHelper' occurred.
Exception Message:
The LHE header is not valid! Weight information was not properly parsed. The error type is 'Header info out of order'

   NOTE: if you want to attempt to process this sample anyway, set failIfInvalidXML = False in the configuration file (current value is True)
.If you set this flag and the error persists, the issue  is fatal and must be solved at the LHE/gridpack level.
```

**Comment by vlimant** [comment]
please test

**Comment by swertz** [comment]
Hi @kdlong , kind ping about the test failures...

File: GeneratorInterface/Core/src/WeightHelper.cc [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+        return true;
+      }
+    }
+    return false;
+  }
+
+  void WeightHelper::cleanupOrphanCentralWeight(WeightGroupInfoContainer& weightGroups) const {
+    auto centralIt = std::find_if(std::begin(weightGroups), std::end(weightGroups), [](auto& entry) {
+      return entry->weightType() == gen::WeightType::kScaleWeights &&
+             static_cast<ScaleWeightGroupInfo*>(entry.get())->containsCentralWeight();
```
Comment (kpedro88) [inline GeneratorInterface/Core/src/WeightHelper.cc:195]: I think the logic here is incorrect: this should select for `!containsCentralWeight()`, in order to find the scale weight group that is *missing* the central weight (because it was orphaned). (I had to change this when testing privately, or the resulting scale weight group did not qualify as well formed.)

**Review by kpedro88 — COMMENTED** [review]
Hi @kdlong, I was looking at this again for private use (since the gen weight situation in UL is even messier than pre-L...) and noticed a few bugs and typos. Just leaving these notes here for whenever you have time to get back to it.

File: GeneratorInterface/Core/src/LHEWeightHelper.cc [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    return curLevel == 0;
+  }
+
+  void LHEWeightHelper::swapHeaders(std::vector<std::string>& headerLines) const {
+    int curLevel = 0;
+    int open = -1;
+    int close = -1;
+    for (size_t idx = 0; idx < headerLines.size(); idx++) {
+      std::string& line = headerLines[idx];
+      std::cout << "Line is " << line << std::endl;
```
Comment (kpedro88) [inline GeneratorInterface/Core/src/LHEWeightHelper.cc:157]: Should be behind `debug_` flag

**Comment by clacaputo** [comment]
-reconstruction

cleaning reco queue

**Comment by smuzaffar** [comment]
Milestone for this pull request has been moved to CMSSW_14_0_X.Please open a backport if it should also go in to CMSSW_13_3_X.

**Comment by vlimant** [comment]
please close
this is definitely stalled. please reopen when necessary

File: PhysicsTools/NanoAOD/plugins/NanoAODOutputModule.cc [hunk 6]
```diff
@@ -299,41 +327,42 @@ void NanoAODOutputModule::openFile(edm::FileBlock const&) {
     m_file->SetCompressionAlgorithm(ROOT::kZLIB);
   } else if (m_compressionAlgorithm == std::string("LZMA")) {
     m_file->SetCompressionAlgorithm(ROOT::kLZMA);
-  } else if (m_compressionAlgorithm == std::string("ZSTD")) {
```
Comment (nsmith-) [inline PhysicsTools/NanoAOD/plugins/NanoAODOutputModule.cc:302]: Is this an artifact of some merge history issue or do we want to remove these compression algos?

Comment (kdlong) [reply PhysicsTools/NanoAOD/plugins/NanoAODOutputModule.cc:302]: That must be a merge mistake, thanks for spotting!
